crates/ra_hir_def/src/nameres/eager.rs

```diff
@@ -0,0 +1,112 @@
+//! Eager expansion collector and resolver
+use hir_expand::{
+    MacroCallId, MacroCallLoc, MacroDefId, MacroEagerCallId, MacroEagerCallLoc, MacroEagerMode,
+    MacroFileKind,
+};
+
+use ra_syntax::ast;
+use std::sync::Arc;
+
+use crate::{
+    db::DefDatabase2,
+    nameres::{raw, CrateDefMap, EagerResolver, ResolveMode},
+    path::{Path, PathKind},
+    AstId, CrateModuleId,
+};
+
+struct EagerMacro {
+    pub def: MacroDefId,
+    pub ast_id: AstId<ast::MacroCall>,
+}
+
+pub(crate) struct DefEagerMacroResolver<'a> {
+    pub def_map: &'a CrateDefMap,
+    pub module_id: CrateModuleId,
+}
+
+impl EagerResolver for DefEagerMacroResolver<'_> {
+    fn resolve_macro(&self, db: &impl DefDatabase2, path: &Path) -> Option<MacroDefId> {
+        // legacy scope
+        if let Some(macro_def) = path
+            .as_ident()
+            .and_then(|name| self.def_map[self.module_id].scope.get_legacy_macro(&name))
+        {
+            return Some(macro_def);
+        }
+
+        let mut path = path.clone();
+        if path.is_ident() {
+            path.kind = PathKind::Self_;
+        }
+
+        // Resolve in module scope, expand during name resolution.
+        let resolved_res =
+            self.def_map.resolve_path_fp_with_macro(db, ResolveMode::Other, self.module_id, &path);
+
+        resolved_res.resolved_def.get_macros()
+    }
+}
+
+/// Collect eager expansion id recursively
+///
+/// Steps:
+/// 1. Expand current eager expansion expand shallowly (in `expand_eager_macros`)
+/// 2. Collect all inner macro calls.
+/// 3. Resolve and expand each inner macro calls recursively
+/// 4. Add all inner macro calls eager id to current eager loc
+/// 5. Return curr eager id
+///
+/// Basically, the final eager call id is the root of tree of eager calls.
+/// In the actual expansion phase, we could rebuild this graph and use it to expand
+/// inner macro calls.
+fn collect_eager_macro_recur(
+    resolver: &impl EagerResolver,
+    db: &impl DefDatabase2,
+    curr: &EagerMacro,
+) -> Option<(MacroEagerCallId)> {
+    let call_id = db.intern_macro(MacroCallLoc {
+        def: curr.def,
+        ast_id: curr.ast_id,
+        eager: Some(MacroEagerMode::Shallow),
+    });
+
+    let file_id = call_id.as_file(MacroFileKind::Expr);
+    let raw_items = db.raw_items(file_id);
```

Comment:
Hm, the fact that we need to build raw_items here seems highly suspicious. 

>Eagerly expanded macros don't add anything to the module structure

So, we shouldn't have code that tries to handle items declared by eager macros, but `raw_items` does exactly that. 

In general, I am highly suspicious about plugging eager expansion into the query system. It seems like it just can't work, because, unlike the case with lazy macros, we can't condence all relevant nameres info to a compact `MacroDefId`.

Let's try do this outside of the query system? I think this could work as follows

1. We have this function

```rust
fn eager_expand(
    macro: EagerMacro,
    arg: tt::Subtree,
    macro_resolver: &dyn Fn(ast::Path) -> Option<UsualOrEagerMacroDef>,
) -> Option<tt::Subtree> {

}
```

it parses `arg` as an `ast::Expe`, and then recursively expands all macros withing, using passed-in `macro_resolver`. Note that, if the `arg` contains a usual macro, it's expansion should also bypass the query infrastructure. That is, this function should *not* accept a `db` argument. 

2. We don't create `MacroFile` and `HIrFileId` for eager macros at all. (That makes it hard to look at the expansion of an eager macro, but I think this can be handled, in the future, by storing such expansions in a table withing `CrateDefMap`).

3. In `nameres::collector`, we special case the `include!` macro, as it's the only eager macro that can affect the set of itmes. Becuse include fundamentally includes a file path, we can use `HirFileId::FileId` to refer to it's contents, we don't need to really identify the result of the exapnsion of `include!` itself. 

4. In expressions, when we see something like `concat!`, we just immediatelly lower the expansion to AST.

Reply:
This probably needs to be even simpler: all eager macros expand to strings (with the expaction of `include` which doesn't expand to syntax at all), so we should have something like rustc's `expr_to_spanned_string`.

Reply:
I think your approach is much simpler than mine. I try to think about it and how to work with `include_str!` : 
```rust
fn main() {
    let _a = concat!(include_str!("test01.txt"), include_str!("test02.txt"));
}
```

Maybe instead of passing a single resolver function, we pass in a trait : 

```rust
trait EagerResolver { 
  fn resolve_macro(path: ast::Path) -> Option<MacroDefId>;
  fn resolve_file(file_name: &str) -> Option<FileId>;
}
```

Reply:
Why would it need `resolve_file`? I think `include!` should be handled *outside* of this function.

Reply:
And I just tried these too and it works too: 
```
/// test01.txt
"Hello"
/// test02.txt
"World"

fn main() {
    let _a = concat!(include!("test01.txt"), include!("test02.txt"));
}
```

Reply:
> Why would it need resolve_file? I think include! should be handled outside of this function.

It is because we need the content of the file to `concat` it inside the `eager_expand`. And personally I think : 
```rust
fn eager_macro(
  macro_call: ast::MacroCall,
  def: MacroDefId,
  macro_resolver: &dyn EagerResolver,
) -> Option<Either<SyntaxNode, FileId>> {   
}
```
It could be easily to use in different context. (hir, collector, impl_block etc.)